docs/dotnet/dotnet-kestrel-metrics.md

```diff
@@ -0,0 +1,221 @@
+# Semantic Conventions for Kestrel web server metrics
+
+**Status**: [Experimental][DocumentStatus]
+
+This document defines semantic conventions for Kestrel web server.
+
+<!-- toc -->
+
+- [Metric: `kestrel.active_connections`](#metric-kestrelactive_connections)
+- [Metric: `kestrel.connection.duration`](#metric-kestrelconnectionduration)
+- [Metric: `kestrel.rejected_connections`](#metric-kestrelrejected_connections)
+- [Metric: `kestrel.queued_connections`](#metric-kestrelqueued_connections)
+- [Metric: `kestrel.queued_requests`](#metric-kestrelqueued_requests)
+- [Metric: `kestrel.upgraded_connections`](#metric-kestrelupgraded_connections)
+- [Metric: `kestrel.tls_handshake.duration`](#metric-kestreltls_handshakeduration)
+- [Metric: `kestrel.active_tls_handshakes`](#metric-kestrelactive_tls_handshakes)
+
+<!-- tocstop -->
+
+## Metric: `kestrel.active_connections`
+
+**Status**: [Experimental][DocumentStatus]
+
+This metric is required.
+
+<!-- semconv metric.kestrel.active_connections(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `kestrel.active_connections` | UpDownCounter | `{connection}` | Number of connections that are currently active on the server. [1] |
+
+**[1]:** Corresponding `EventCounter` name is `current-connections`; Meter name is `Microsoft.AspNetCore.Server.Kestrel`
+<!-- endsemconv -->
+
+<!-- semconv metric.kestrel.active_connections(full) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `kestrel.endpoint` | string | TODO [1] | `TODO` | Recommended |
+
+**[1]:** **TODO: why not server.**? What's KEstrel-specific about it**
+<!-- endsemconv -->
+
+## Metric: `kestrel.connection.duration`
+
+**Status**: [Experimental][DocumentStatus]
+
+This metric is required.
+
+<!-- semconv metric.kestrel.connection.duration(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `kestrel.connection.duration` | Histogram | `s` | The duration of connections on the server. [1] |
+
+**[1]:** Meter name is `Microsoft.AspNetCore.Server.Kestrel`
+<!-- endsemconv -->
+
+<!-- semconv metric.kestrel.connection.duration(full) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `exception.type` | string | The type of the exception (its fully-qualified class name, if applicable). The dynamic type of the exception should be preferred over the static type in languages that support it. [1] | `java.net.ConnectException`; `OSError` | Recommended |
+| `kestrel.endpoint` | string | TODO [2] | `TODO` | Recommended |
+| [`network.protocol.version`](../general/attributes.md) | string | Version of the application layer protocol used. See note below. [3] | `3.1.1` | Recommended |
+| `tls.protocol` | string | TODO [4] | `TODO` | Recommended |
```

Comment:
Thanks this is a very useful reference! 

In addition to tls.version defined there I also see tls.version_protocol. How would folks feel if we conditionally used tls.version_protocol only when protocol != "tls"? SSL is in .NET's internal enumeration so I assume it technically could get used, but it is also quite outdated and marked obsolete so probably gets used rarely. It seems annoying to be sending an extra attribute that is nearly always "tls" when we could just assume it is "tls" unless told otherwise.